### PR TITLE
Importing all the updates from the LOTR version of the bot (except spreadsheets)

### PR DESCRIPTION
**Pull request by ketura**
This includes all of the improvements done over the last year or so to the LOTR version of this bot, *except* for the spreadsheet rework, which was objected to last time.  These changes include:

1. Expanded the input and key scrubbing so that all non-alphanumeric characters are ignored.  This makes anything short of a literal misspelling work (spacing and punctuation errors now have no effect)
2. Removed caching support (it "locked-in" user selections into the cache, which isn't great for situations where there are multiple legit alternatives, which conflicts with the search support below)
3. Combined all query options together instead of short-circuiting out if a higher priority tier is found (to ensure that all alternatives are found)
4. Added fuzzy-ish search at the end of the query process to search for cards containing the query rather than exactly matching it (if the query is longer than 2 characters).
5. Overhauled bot to utilize buttons and dropdowns instead of reactions.  Dropdown can only be used by the person who summoned the bot or the server owner (which is the same behavior as the reactions).  Anyone can open the dropdown to browse the results, however.
6. Bot responses automatically include a "delete" button which can be used to eliminate unwanted responses (tho only by the person who summoned it or the server owner)
7. Bot responses automatically include a "wiki" button which links unobtrusively to the wiki link
8. Added support for the bot responding in Discord threads
9. Added check to ignore any spreadsheet entries that have a blank ID or coll info (as a proxy for skipping blank lines in the spreadsheet)
10. Added column trimming, since a bunch of the entries have random spaces in them
11. Temporarily disabled loading from the nicknames field as the current iteration of the spreadsheet results in tens of thousands of redundant search parameters which slows down the bot, almost none of which are actually increasing usability.  The useful nicks are now covered by the input scrubbing typo compensation + dynamic drop-down switching.
12. Added handling for a 25-card limit in the response (the max that a dropdown can hold, and coincidentally the maximum number of reacts that could have been used on the old bot anyway)
13. Altered the summoning syntax to [[card name]], to cut down dramatically on false positive responses (and to match familiar syntax for other popular bots)
14. Removed {{card name}} syntax, as the image response now always includes a button that will link to the wiki

----

Gallery of major change results (ignore the bot name, I lazily used the LOTR bot for testing):

Abbreviations of cards with and without punctuation:
image

The drop-down when multiple cards are found:
image

After selecting an image, the dropdown doesn't go away (until the "accept" button is clicked), allowing one to re-select an option when searching for a particular variant:
image

A slightly different message when more than 25 cards are found:
image

Delete button on failed searches.  Also shows the split for cards that have (sub-names in parentheses) in the title; both are now searchable: 
image

No amount of space or punctuation errors get in the way, and as you can see all of the search results are relevant (none of the options not shown had anything except "obi-wan" in the title):
image

Demonstrating the use of search when only part of a title is matched:
image

### CODE REVIEW

File: DiscordCardLinker/CardBot.cs [hunk 1]
```diff
@@ -130,19 +160,19 @@ public class CardBot {
 			 * Only add the card key once.
 			 */
 			if(!collection.ContainsKey(key)) {
-				Console.WriteLine("Adding Key: ["+key+"]");
+				//Console.WriteLine("Adding Key: ["+key+"]");
```
Comment (DevoKun) [inline DiscordCardLinker/CardBot.cs:163]: Please do not comment out debugging information. This information is critical for debugging the container service if it is in a frozen state.

File: DiscordCardLinker/CardBot.cs [hunk 2]
```diff
@@ -130,19 +160,19 @@ public class CardBot {
...
-			Console.WriteLine("  .. ID.........: "+card.ID);
-			Console.WriteLine("  .. ImageURL...: "+card.ImageURL);
-			Console.WriteLine("  .. WikiURL....: "+card.WikiURL);
-			Console.WriteLine("  .. CollInfo...: "+card.CollInfo);
-			Console.WriteLine("  .. DisplayName: "+card.DisplayName);
-			Console.WriteLine("  .. Title......: "+card.Title);
-			Console.WriteLine("  .. Subtitle...: "+card.Subtitle);
-			Console.WriteLine("  .. TitleSuffix: "+card.TitleSuffix);
-			Console.WriteLine("  .. Nicknames..: "+card.Nicknames);
+			//Console.WriteLine("  .. ID.........: "+card.ID);
```
Comment (DevoKun) [inline DiscordCardLinker/CardBot.cs:167]: Please do not comment out debugging information. This information is critical for debugging the container service if it is in a frozen state.

File: DiscordCardLinker/CardBot.cs [hunk 3]
```diff
@@ -154,103 +184,98 @@ public class CardBot {
...
+			Client.ThreadUpdated += OnThreadUpdated;
 
 			// https://dsharpplus.github.io/api/DSharpPlus.DiscordClient.html?q=ConnectAsync#DSharpPlus_DiscordClient_ConnectAsync_DiscordActivity_System_Nullable_UserStatus__System_Nullable_DateTimeOffset__
 			/* connecting to discord */
 			await Client.ConnectAsync();
 			/* successfully connected to discord */
 		}
 
-		/*
-		 * process reaction to discord chat message
```
Comment (DevoKun) [inline DiscordCardLinker/CardBot.cs:166]: Please do not remove comments describing what the code does.

File: DiscordCardLinker/CardBot.cs [hunk 4]
```diff
@@ -311,166 +318,123 @@ public class CardBot {
...
+		{
+			var builder = new DiscordMessageBuilder()
+				.WithReply(original.Id)
+				.WithContent(card.ImageURL);
+
+			var comps = new List<DiscordComponent>();
+
+			if (wiki)
+			{
+				comps.Add(new DiscordLinkButtonComponent(card.WikiURL, "Wiki", false));
```
Comment (DevoKun) [inline DiscordCardLinker/CardBot.cs:387]: SWCCG does not have a "Wiki". This button should say "scomp"

**Review by DevoKun — CHANGES_REQUESTED** [review]

File: DiscordCardLinker/CardBot.cs (hunk 1 above)
Comment (ketura) [reply DiscordCardLinker/CardBot.cs:163]: Apologies, forgot to restore this (it was causing a delay of a few minutes each run).

File: DiscordCardLinker/CardBot.cs (hunk 4 above)
Comment (ketura) [reply DiscordCardLinker/CardBot.cs:387]: Fair enough.

File: DiscordCardLinker/CardBot.cs (hunk 3 above)
Comment (ketura) [reply DiscordCardLinker/CardBot.cs:166]: That function was removed and so the orphan comment with it.  I can add comments to the currently uncommented functions.

**Review by DevoKun — APPROVED** [review]

**Pull request merged**